### PR TITLE
Add How does ML.NET work?

### PR DESCRIPTION
**Pull request by natke**
## Summary

Add new article: https://review.docs.microsoft.com/en-us/dotnet/machine-learning/how-does-mldotnet-work?branch=pr-en-us-11871

[AB#1479484](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1479484)

Note: art work is preliminary but please give feedback on its content.

ETA: there are currently two broken links in this PR to the load data and train model how-tos. I'm assuming that @luisquintanilla's PRs will get merged before this one.

### CODE REVIEW

**Review by luisquintanilla — CHANGES_REQUESTED** [review]

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+The simplest model is two-dimensional linear regression, where one continuous quantity is proportional to another, as in the house price example above. 
+
+![Linear Regression Model with bias and weight parameters](./media/linear-regression-model.svg)
+
+The model is simply: `Price = b + Size * w`. The parameters `b` and `w` are estimated by fitting a line on a set of (size, price) pairs. The data used to find the parameters of the model is called **training data**. The inputs of a machine learning model are called **features**. In this example, `Size` is the only feature. The ground-truth values used to train a machine learning model are called **labels**. Here, the `Price` values in the training data set are the labels. 
+
+### More complex
+
+A more complex model classifies financial transactions into categories using the text description of the transaction.
```
Comment (JRAlexander) [inline docs/machine-learning/how-does-mldotnet-work.md]: A more complex model classifies financial transactions into categories using the transaction text description.

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+Each transaction description is broken down into a set of features by removing redundant words and characters, counting word and character combinations, then training a linear model based on the set of categories in the training data. The more similar a new description is to the ones in the training set, the more likely it will be assigned to the same category.
```
Comment (JRAlexander) [inline docs/machine-learning/how-does-mldotnet-work.md]: Maybe something like: "Each transaction description is broken down into a set of features by removing redundant words and characters, and counting word and character combinations. The feature set trains a linear model based on the set of categories in the training data."

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+        var metrics = mlContext.Regression.Evaluate(testPriceDataView, labelColumnName: "Price");
+
+        Console.WriteLine($"R^2: {metrics.RSquared:0.##}");
+        Console.WriteLine($"RMS error: {metrics.RootMeanSquaredError:0.##}");
+
+        // R^2: 0.99
+        // RMS error: 0.09
+```
+
+The evaluation metrics tell us that the error is low-ish, and that correlation between the predicted output and the test output is high. That was easy! In real examples, it takes more tuning to achieve good model metrics.
```
Comment (JRAlexander) [inline docs/machine-learning/how-does-mldotnet-work.md]: "us" -> "you". Use second person.

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+## Architecture
+
+In this section, we go through the architectural patterns of ML.NET. If you are an experienced .NET developer, some of these patterns will be familiar to you, and some will be less familiar. Hold tight, while we dive in!
```
[the remark this answers is not in the document]
Comment (JRAlexander) [reply docs/machine-learning/how-does-mldotnet-work.md]: "we" -> "you".

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+```csharp
+    var pipeline = mlContext.Transforms.Concatenate("Features", new[] { "Size" })
+        .Append(mlContext.Regression.Trainers.Sdca(labelColumnName: "Price", maximumNumberOfIterations: 100);
+```
+
+In the snippet, `Concatenate` and `Sdca` are both extension functions. They each create an <xref:Microsoft.ML.IEstimator`1> object that is appended to the pipeline. At this point, the objects are created only. No execution has happened.
+
+### Fit the model
+
+Once the objects in the pipeline have been created, data can be used to train or fit the model.
```
Comment (JRAlexander) [inline docs/machine-learning/how-does-mldotnet-work.md]: Would this be better? "Once the objects in the pipeline have been created, data can be used to fit (train) the model."

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+
+```csharp
+    var model = pipeline.Fit(trainingData);
+```
+
+Calling `Fit()` uses the input training data to estimate the parameters of the model. Remember in our linear regression model above, we had two model parameters: **bias** and **weight**. After the `Fit()` call, the values of the parameters are known. Most models will have many more parameters than this.
```
[the remark this answers is not in the document]
Comment (JRAlexander) [reply docs/machine-learning/how-does-mldotnet-work.md]: Maybe something like this: "Calling `Fit()` uses the input training data to estimate the parameters of the model. This is known as training the model."

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+   ![ML.NET Data View example with house price prediction data](./media/ml-net-dataview.png)
+
+All ML.NET algorithms look for an input column that is a vector. By default this vector column is called **Features**. This is why we concatenated the **Size** column into a new column called **Features** in our house price example.
+
+ ```csharp
+    var pipeline = mlContext.Transforms.Concatenate("Features", new[] { "Size" })
+ ```
+
+All algorithms also create new columns after they have performed a prediction. The fixed names of these new columns depend on the type of machine learning algorithm. For the regression task, one of the new columns is called **Score**. This is why we attributed our price data with this name.
```
Comment (JRAlexander) [inline docs/machine-learning/how-does-mldotnet-work.md]: "This is why you attributed your price data with this name."

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+
+You can find out more about output columns of different machine learning tasks in the [Machine Learning Tasks](resources/tasks.md) guide.
+
+An important property of DataView objects is that they are evaluated **lazily**. Data views are only loaded and operated on during model training and evaluation, and data prediction. While you are writing and testing your ML.NET application, you can use the Visual Studio debugger to take a peek at any data view object by calling the [Preview](xref:Microsoft.ML.DebuggerExtensions.Preview*) method.
+
+```csharp
+    var debug = testPriceDataView.Preview();
+```
+
+You can watch the `debug` variable in the debugger and examine its contents. Do not use the Preview method in production code, as it significantly degrades performance.
```
Comment (JRAlexander) [inline docs/machine-learning/how-does-mldotnet-work.md]: "Do not" -> "Don't"

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+
+### More complex
+
+A more complex model classifies financial transactions into categories using the text description of the transaction.
+
+Each transaction description is broken down into a set of features by removing redundant words and characters, counting word and character combinations, then training a linear model based on the set of categories in the training data. The more similar a new description is to the ones in the training set, the more likely it will be assigned to the same category. 
+
+![Text Classification Model](./media/text-classification-model.png)
+
+Both the house price model and the text classification model are **linear** models. Depending on the nature of your data and the problem you are solving, you can also use **decision tree** models, **generalized additive** models and others. You can find out more about the models in [How to choose an ML.NET model](./resources/tasks.md).
```
Comment (JRAlexander) [inline docs/machine-learning/how-does-mldotnet-work.md]: "you are" -> "you're" and there should be a comma after "**generalized additive** models".

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+
+## Data preparation
+
+In most cases, the data that you have available in your application is not suitable to be used directly to train a machine learning model. The raw data needs to be prepared, or pre-processed before it can be used to find the parameters of your model. Your data may need to be converted from string values to a numerical representation. You might have redundant information in your input data. You may need to reduce or expand the dimensions of your input data. Your data might need to be normalized or scaled.
```
Comment (JRAlexander) [inline docs/machine-learning/how-does-mldotnet-work.md]: "the data that you have available in your application " -> "your application data"

Comment (JRAlexander) [reply docs/machine-learning/how-does-mldotnet-work.md]: "is not" -> "isn't"

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+ 
+The `CreatePredictionEngine()` method takes an input class and an output class. The field names and/or code attributes determine the names of the data columns used during model training and prediction. You can read about  [How to make a single prediction](./how-to-guides/single-predict-model-ml-net.md) in the How-to section.
+
+### Data models and schema
+
+At the core of an ML.NET machine learning pipeline are [DataView](xref:Microsoft.ML.IDataView) objects.
+
+Each transformation in the pipeline has an input schema (data names, types, and sizes that the transform expects to see on its input); and an output schema (data names, types, and sizes that the transform produces after the transformation). 
+
+If the output schema from one transform in the pipeline does not match the input schema of the next transform, ML.NET will throw an exception.
```
Comment (JRAlexander) [inline docs/machine-learning/how-does-mldotnet-work.md]: "does not" > "doesn't"

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+
+A data view object has columns and rows. Each column has a name and a type and a length. For example: the input columns in the house price example are **Size** and **Price**. They are both type <xref:System.Single> and they are scalar quantities rather than vector ones.
```
[the remark this answers is not in the document]
Comment (JRAlexander) [reply docs/machine-learning/how-does-mldotnet-work.md]: "Each column name has a name, type, and length"

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+
+   ![ML.NET Data View example with house price prediction data](./media/ml-net-dataview.png)
+
+All ML.NET algorithms look for an input column that is a vector. By default this vector column is called **Features**. This is why we concatenated the **Size** column into a new column called **Features** in our house price example.
```
Comment (JRAlexander) [inline docs/machine-learning/how-does-mldotnet-work.md]: "This is why the **Size** column was concatenated into a new column called **Features** in the house price example."

**Review by JRAlexander — CHANGES_REQUESTED** [review]
Looks good. A few small items.

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+
+## Architecture
+
+In this section, we go through the architectural patterns of ML.NET. If you are an experienced .NET developer, some of these patterns will be familiar to you, and some will be less familiar. Hold tight, while we dive in!
+
+An ML.NET application starts with an <xref:Microsoft.ML.MLContext> object. This singleton object contains properties, which are **catalogs** for:
+
+- data loading and saving: <xref:Microsoft.ML.Data>
+- data preparation :<xref:Microsoft.ML.TransformsCatalog>
+- algorithms for model training
```
[the remark this answers is not in the document]
Comment (natke) [reply docs/machine-learning/how-does-mldotnet-work.md]: Where would you link to here @luisquintanilla?

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+
+```csharp
+    var model = pipeline.Fit(trainingData);
+```
+
+Calling `Fit()` uses the input training data to estimate the parameters of the model. Remember in our linear regression model above, we had two model parameters: **bias** and **weight**. After the `Fit()` call, the values of the parameters are known. Most models will have many more parameters than this.
+
+You can learn more about model training in [How to train your model](./how-to-guides/train-model.md)
+
+The resulting model object implements the <xref:Microsoft.ML.ITransformer> interface. That is, the model transforms transforms input data into predictions.
```
[the remark this answers is not in the document]
Comment (natke) [reply docs/machine-learning/how-does-mldotnet-work.md]: Good catch, that's a typo

File: docs/machine-learning/how-does-mldotnet-work.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (natke) [reply docs/machine-learning/how-does-mldotnet-work.md]: ?

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+    public class Prediction
+    {
+        [ColumnName("Score")]
+        public float Price { get; set; }
+    }
+```    
+
+You can find out more about output columns of different machine learning tasks in the [Machine Learning Tasks](resources/tasks.md) guide.
+
+An important property of DataView objects is that they are evaluated **lazily**. Data views are only loaded and operated on during model training and evaluation, and data prediction. While you are writing and testing your ML.NET application, you can use the Visual Studio debugger to take a peek at any data view object by calling the [Preview](xref:Microsoft.ML.DebuggerExtensions.Preview*) method.
```
[the remark this answers is not in the document]
Comment (natke) [reply docs/machine-learning/how-does-mldotnet-work.md]: I was actually trying to group model training and evaluation, and keep prediction separate. Any other suggestions to keep this intent?

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+#Customer intent: As a developer, I want to learn how ML.NET works so that I can leverage machine learning in my applications.
+---
+
+# How does ML.NET work?
+
+ML.NET gives you the ability to add machine learning to your .NET application. With this capability, you can make automatic predictions using the data available to your application.
+
+Examples of the type of prediction that you can make with ML.NET include:
+
+- Categorization, or classification. For example: automatically divide customer feedback into positive and negative categories
```
[the remark this answers is not in the document]
Comment (natke) [reply docs/machine-learning/how-does-mldotnet-work.md]: Great suggestion!

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+
+An ML.NET model is an encapsulation of the transformations to perform on your input data to arrive at the predicted output.
+
+### Simple
+
+The simplest model is two-dimensional linear regression, where one continuous quantity is proportional to another, as in the house price example above. 
+
+![Linear Regression Model with bias and weight parameters](./media/linear-regression-model.svg)
+
+The model is simply: `Price = b + Size * w`. The parameters `b` and `w` are estimated by fitting a line on a set of (size, price) pairs. The data used to find the parameters of the model is called **training data**. The inputs of a machine learning model are called **features**. In this example, `Size` is the only feature. The ground-truth values used to train a machine learning model are called **labels**. Here, the `Price` values in the training data set are the labels.
```
[the remark this answers is not in the document]
Comment (natke) [reply docs/machine-learning/how-does-mldotnet-work.md]: In this case, this is what the model is. Do you think it is confusing?

File: docs/machine-learning/how-does-mldotnet-work.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (natke) [reply docs/machine-learning/how-does-mldotnet-work.md]: Good suggestion

File: docs/machine-learning/how-does-mldotnet-work.md (hunk 5 above)
Comment (natke) [reply docs/machine-learning/how-does-mldotnet-work.md]: Removed fit altogether

File: docs/machine-learning/how-does-mldotnet-work.md (hunk 10 above)
Comment (natke) [reply docs/machine-learning/how-does-mldotnet-work.md]: removed in your application. Application data sounds more like environment variables and the like

**Comment by natke** [comment]
@JRAlexander and @luisquintanilla, I've updated after review. I addressed *most* of the comments with your suggestions. Some I didn't. We can discuss if you like. Can you please re-review?

There is one conceptual diagram to come: the code workflow one.

File: docs/machine-learning/how-does-mldotnet-work.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply docs/machine-learning/how-does-mldotnet-work.md]: Hey, saw you made the change. The links you used were the ones I was referring to.

**Review by luisquintanilla — APPROVED** [review]

File: docs/machine-learning/how-does-mldotnet-work.md [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+    var debug = testPriceDataView.Preview();
+```
+
+You can watch the `debug` variable in the debugger and examine its contents. Do not use the Preview method in production code, as it significantly degrades performance.
+
+### Model Deployment
+
+In real-life applications, your model training and evaluation code will be separate from your prediction. In fact, these two activities are often performed by separate teams. Your model development team can save the model for use in the prediction application.
+
+```csharp
```
Comment (luisquintanilla) [inline docs/machine-learning/how-does-mldotnet-work.md]: Suggestion: This snippet can be simplified since the model is being saved to a local file. Use the method that takes in the string path as opposed to a Stream.

```csharp
mlContext.Model.Save(model, trainingData.Schema,"model.zip");
```

**Review by JRAlexander — APPROVED** [review]

File: docs/machine-learning/how-does-mldotnet-work.md (hunk 19 above)
Comment (natke) [reply docs/machine-learning/how-does-mldotnet-work.md]: Well-spotted. Fixing, thanks!
